atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/persistentlock/PersistentLock.java

```diff
@@ -0,0 +1,161 @@
+        return acquireLock(lockName, reason, true);
+    }
+
+    public LockEntry acquireLock(
+            PersistentLockName lockName, String reason, boolean exclusive) throws PersistentLockIsTakenException {
+        long thisId = ThreadLocalRandom.current().nextLong(Long.MAX_VALUE);
+        LockEntry lockEntry = LockEntry.of(lockName, thisId, reason, exclusive);
+
+        insertLockEntry(lockEntry);
+
+        return verifyLockWasAcquired(lockEntry);
+    }
+
+    public void releaseLock(LockEntry lock) {
+        log.debug("Releasing persistent lock " + lock);
+        keyValueService.delete(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lock.deletionMapWithTimestamp(LOCKS_TIMESTAMP));
+    }
+
+    private LockEntry verifyLockWasAcquired(LockEntry desiredLock) throws PersistentLockIsTakenException {
+        List<LockEntry> relevantLocks = allRelevantLockEntries(desiredLock.lockName());
+        Preconditions.checkState(relevantLocks.contains(desiredLock), "Lock was not properly inserted");
+
+        List<LockEntry> conflictingLocks = removeSingleLock(desiredLock, relevantLocks);
+
+        if (!desiredLock.exclusive()) {
+            conflictingLocks = retainExclusiveLocks(conflictingLocks);
+        }
+        return verifyLockDoesNotConflict(desiredLock, conflictingLocks);
+    }
+
+    private List<LockEntry> retainExclusiveLocks(List<LockEntry> conflictingLocks) {
+        return conflictingLocks.stream()
+                .filter(LockEntry::exclusive)
+                .collect(Collectors.toList());
+    }
+
+    private List<LockEntry> removeSingleLock(LockEntry desiredLock, List<LockEntry> relevantLocks) {
+        return relevantLocks.stream()
+                .filter(otherLock -> !otherLock.equals(desiredLock))
+                .collect(Collectors.toList());
+    }
+
+    private LockEntry verifyLockDoesNotConflict(
+            LockEntry desiredLock,
+            List<LockEntry> conflictingLocks) throws PersistentLockIsTakenException {
+
+        if (conflictingLocks.isEmpty()) {
+            log.debug("Acquired persistent lock " + desiredLock);
+            return desiredLock;
+        } else {
+            log.info("Failed to acquire persistent lock " + desiredLock);
+            releaseLock(desiredLock);
+            throw new PersistentLockIsTakenException(conflictingLocks);
+        }
+    }
+
+    private void insertLockEntry(LockEntry lockEntry) {
+        log.debug("Attempting to acquire persistent lock " + lockEntry);
+        keyValueService.put(AtlasDbConstants.PERSISTED_LOCKS_TABLE, lockEntry.insertionMap(), LOCKS_TIMESTAMP);
+    }
+
+    private List<LockEntry> allRelevantLockEntries(PersistentLockName lock) {
+        return allLockEntries().stream()
+                .filter(lockEntry -> lockEntry.lockName().equals(lock))
+                .collect(Collectors.toList());
+    }
+
+    public List<LockEntry> allLockEntries() {
+        List<RowResult<Value>> allLockRows = ImmutableList.copyOf(keyValueService.getRange(
```

Comment:
I envision this table containing very few entries at any time, and I find this to be more readable.